lib/smf/smf.c

```diff
@@ -0,0 +1,319 @@
+ */
+static bool smf_execute_ancestor_run_actions(struct smf_ctx *ctx)
+{
+	struct internal_ctx * const internal = (void *) &ctx->internal;
+	const struct smf_state *tmp_state;
+
+	/* Execute all run actions in reverse order */
+
+	/* Return if the current state switched states */
+	if (internal->new_state) {
+		internal->new_state = false;
+		return false;
+	}
+
+	/* Return if the current state terminated */
+	if (internal->terminate) {
+		return true;
+	}
+
+	/* Try to run parent run actions */
+	tmp_state = ctx->current->parent;
+	for (int i = 0; i < CONFIG_NUM_SMF_ANCESTORS; i++) {
+		/* Execute parent run action */
+		if (tmp_state && tmp_state->run) {
+			tmp_state->run(ctx);
+		}
+
+		/* break if the parent state switched states */
+		if (internal->new_state) {
+			break;
+		}
+
+		/* No need to continue if terminate was set */
+		if (internal->terminate) {
+			return true;
+		}
+
+		/* Get next parent state */
+		if (tmp_state) {
+			tmp_state = tmp_state->parent;
+		}
+	}
+
+	internal->new_state = false;
+	/* All done executing the run actions */
+
+	return false;
+}
+
+/**
+ * @brief Execute all ancestor exit actions
+ *
+ * @param ctx State machine context
+ * @param target The exit actions of this target's ancestors are executed
+ * @return true if the state machine should terminate, else false
+ */
+static bool smf_execute_ancestor_exit_actions(struct smf_ctx *const ctx,
+					const struct smf_state *target)
+{
+	struct internal_ctx * const internal = (void *) &ctx->internal;
+	const struct smf_state *tmp_state;
+	const struct smf_state *target_parent;
+
+	/* Execute all parent exit actions in reverse order */
+
+	/* Get target state's parent state */
+	target_parent = target->parent;
+	tmp_state = ctx->current;
+
+	for (int i = 0; i < CONFIG_NUM_SMF_ANCESTORS; i++) {
+		/* Get parent state */
+		if (tmp_state == NULL) {
+			break;
+		}
+
+		tmp_state = tmp_state->parent;
+
+		/*
+		 * Do not execute a parent state's exit action that has
+		 * a shared ancestry with the target.
+		 */
+		while (target_parent != NULL) {
+			if (tmp_state == target_parent) {
+				tmp_state = NULL;
+				break;
+			}
+
+			/* Get target state next parent state if it exists */
+			target_parent = target_parent->parent;
+		}
+
+		if (tmp_state && tmp_state->exit) {
+			/* Execute exit action */
+			tmp_state->exit(ctx);
+
+			/* No need to continue if terminate was set */
+			if (internal->terminate) {
+				return true;
+			}
+		}
+	}
+	return false;
+}
+#endif
+
+void smf_set_initial(struct smf_ctx *ctx, const struct smf_state *init_state)
+{
+	struct internal_ctx * const internal = (void *) &ctx->internal;
+
+	internal->exit = false;
+	internal->terminate = false;
+	ctx->current = init_state;
+	ctx->previous = NULL;
+	ctx->terminate_val = 0;
+
+#ifdef CONFIG_SMF_ANCESTOR_SUPPORT
+	internal->new_state = false;
+
+	if (smf_execute_ancestor_entry_actions(ctx, init_state)) {
+		return;
+	}
+#endif
```

Comment:
See previous comment,  if new_state is allowed when CONFIG_SMF_ANCESTOR_SUPPORT=n, then this can use IS_ENABLED(), which is the preferred style when possible.

```
if (IS_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT)) {
    internal->new_state = false;
    if (smf_execute_ancestor...)
```